src/cmdline/mod.rs

```diff
@@ -254,12 +199,18 @@ impl Cmdline {
     ///
     /// ```rust
     /// # use linux_loader::cmdline::*;
-    /// let mut cl = Cmdline::new(10);
-    /// cl.insert_str("foobar");
+    /// let mut cl = Cmdline::new(100);
+    /// assert!(cl.insert_boot_arg("foobar").is_ok());
     /// assert_eq!(cl.as_str(), "foobar");
     /// ```
-    pub fn as_str(&self) -> &str {
-        self.line.as_str()
+    pub fn as_str(&self) -> String {
+        if self.init_args.is_empty() {
+            self.boot_args.to_string()
+        } else if self.boot_args.is_empty() {
+            format!("-- {}", self.init_args)
```

Comment:
Would this be a valid kernel command line? I wonder if we should return an error in this case.

Reply:
Regarding String::with_capacity(cmdline_capacity) I propose to declare it as String::new() and not provide any initial capacity at all. Before any insertion, we perform capacity checks considering both boot args and init args (if exists) such that insertion of too large args to be impossible. Do you have any concern with this approach?

Reply:
Regarding String::with_capacity(cmdline_capacity) I propose to declare it as String::new() and not provide any initial capacity at all. Before any insertion, we perform capacity checks considering both boot args and init args (if exists) such that insertion of too large args to be impossible. Do you have any concern with this approach?

Reply:
Regarding the usecase in which we return format!("-- {}", self.init_args), I don't see it as an error because the method is called as_str so it does not imply any validity checks done by the cmdline; it just implies a representation of the current state of the cmdline; furthermore, we do not perform any validity checks of the boot arguments the user can insert into the cmdline either (apart from the basic check to make sure it's an ASCII string and it doesn't contain any space); thus, it's the responsibility of the user to provide correct data, we just offer the data structure for implementing a Cmdline.

Reply:
The meaning of `as_str` is that you can pass this to the kernel, so in that sense the `Cmdline` will not get any more pushes. We are working with a fail fast methodology. That means that if there are error cases that we can check ourselves in a reasonable way, we should do that. For example, that is why are checking for ASCII only strings. We could also not check, and let the Linux Kernel fail to boot. But that would mean that we are passing erroneous data down to the lower levels for no good reason. Hence, my question whether Linux accepts a command line that only has init args.

Reply:
Yes, you are right somehow it's always better to catch errors as early as possible. I'll investigate this usecase and come back with a result.